### PR TITLE
Credits/Shop: Add hooks for credit related events.

### PR DESCRIPTION
**Pull request by pkillboredom**
This change adds some hooks for credit related events that are used by my addon [ttt2stats](https://github.com/pkillboredom/ttt2stats). This addon records stats from my server's games to a database, which is used to populate [a dashboard](https://github.com/pkillboredom/TTT2StatsDashboard/tree/master?tab=readme-ov-file). I've been running these changes for a while and its about time I open a PR.

Added:
- Added hook `GM:TTT2OnGiveFoundCredits()`, which is triggered serverside when a player has been given credits for searching a corpse.
- Added hook `GM:TTT2ReceivedKillCredits()`, which is called when a player recieves credits for a kill.
- Added hook `GM:TTT2ReceivedTeamAwardCredits()`, which is called when a player recieves credits as a team award.
- Added hook `GM:TTT2TransferedCredits()`, which is called when a player has successfully transfered a credit to another player.

Changed: 
- TryRerollShop calls `TTT2OrderedEquipment` hook.

LMK if there are any suggestions. The thing I'm iffiest on is the TryRerollShop change, maybe a new hook would be better. Thanks!

### CODE REVIEW

**Review by EntranceJew — APPROVED** [review]

**Review by EntranceJew — APPROVED** [review]

**Comment by pkillboredom** [comment]
Sorry, one more commit. I couldn't understand what the remaining error was in CI so I had to go get stylua.

**Review by EntranceJew — APPROVED** [review]

**Comment by EntranceJew** [comment]
Yeah, it can be a little confusing what it wants, especially if TTT2 isn't the root of your project (for vscode at least)

**Comment by TimGoll** [comment]
I will review this later, but just to ask, are you aware of TTT2's events? I think you should be able to get all of this data already from the events at the round end

**Comment by TimGoll** [comment]
> - Added hook `GM:TTT2OnGiveFoundCredits()`, which is triggered serverside when a player has been given credits for searching a corpse.
> - Added hook `GM:TTT2ReceivedKillCredits()`, which is called when a player recieves credits for a kill.
> - Added hook `GM:TTT2ReceivedTeamAwardCredits()`, which is called when a player recieves credits as a team award.
> - Added hook `GM:TTT2TransferedCredits()`, which is called when a player has successfully transfered a credit to another player.

IMHO these names are inconsistent. It should probably be more like `GM:TTT2OnGiveFoundCredits()`, `GM:TTT2OnReceiveKillCredits()`, `GM:TTT2OnReceiveTeamAwardCredits()` and `GM:TTT2OnTransferCredits()`

File: gamemodes/terrortown/gamemode/server/sv_corpse.lua [hunk 1]
```diff
@@ -732,3 +732,12 @@ function GM:TTT2ModifyRagdollVelocity(deadply, rag, velocity) end
...
 function GM:TTTOnCorpseCreated(rag, deadply) end
+
+---
+-- Called after a player has been given credits for searching a corpse.
+-- @param Player ply The player that searched the corpse
+-- @param Entity rag The ragdoll that was searched
+-- @param number credits The amount of credits that were given
+-- @hook
+-- @realm server
+function GM:TTT2OnGiveFoundCredits(ply, rag, credits) end
```
Comment (TimGoll) [inline gamemodes/terrortown/gamemode/server/sv_corpse.lua:743]: This hook seems to be called no where?

**Review by TimGoll — CHANGES_REQUESTED** [review]
Besides the suggested name change I'd also like to suggest adding more functionality to those hooks. I'm thinking of "return false to cancel the action".

So this
```lua
hook.Add("TTT2OnGiveFoundCredits", "myhook", function(ply, rag, credits)
    return false
end
```
would cancel that action.

But this would probably break your statistics addon, wouldn't it?

**Comment by pkillboredom** [comment]
> I will review this later, but just to ask, are you aware of TTT2's events? I think you should be able to get all of this data already from the events at the round end

In my case I did not want to wait until round end, I wanted to update the database as soon as something had occurred.

> > * Added hook `GM:TTT2OnGiveFoundCredits()`, which is triggered serverside when a player has been given credits for searching a corpse.
> > * Added hook `GM:TTT2ReceivedKillCredits()`, which is called when a player recieves credits for a kill.
> > * Added hook `GM:TTT2ReceivedTeamAwardCredits()`, which is called when a player recieves credits as a team award.
> > * Added hook `GM:TTT2TransferedCredits()`, which is called when a player has successfully transfered a credit to another player.
> 
> IMHO these names are inconsistent. It should probably be more like `GM:TTT2OnGiveFoundCredits()`, `GM:TTT2OnReceiveKillCredits()`, `GM:TTT2OnReceiveTeamAwardCredits()` and `GM:TTT2OnTransferCredits()`

I have changed the names.

File: gamemodes/terrortown/gamemode/server/sv_corpse.lua (hunk 1 above)
Comment (pkillboredom) [reply gamemodes/terrortown/gamemode/server/sv_corpse.lua]: I missed in merging from upstream that the place where this was called moved from sv_corpse to bodysearch. I've added the call back and will test sometime in the next few days.

Comment (pkillboredom) [reply gamemodes/terrortown/gamemode/server/sv_corpse.lua]: I'm not sure that adding the additional functionality you've suggested would work for my stats addon, like you said.

The event system currently doesn't have events for the `GM:TTT2OnReceiveKillCredits()`, `GM:TTT2OnReceiveTeamAwardCredits()` or `GM:TTT2OnTransferCredits()` cases. If you would prefer that I pivot toward using the event system, I think that would work for me over making the hooks cancel-able. I suppose I would just need to add events for those cases and listen to GM:TTT2AddedEvent to record the events in real time, right?

Comment (TimGoll) [reply gamemodes/terrortown/gamemode/server/sv_corpse.lua]: Nah, I don't want to pivot you in that direction, I only wanted to make sure that you are aware of it. If you say that it doesn't work for you, then it is fine

**Review by TimGoll — APPROVED** [review]
This somehow got lost as you never updated the review state and I never checked the code state. LGTM, thanks for contributing!

**Pull request merged**